src/simplewallet/simplewallet.cpp

```diff
@@ -1789,47 +1825,84 @@ bool simple_wallet::refresh(const std::vector<std::string>& args)
   success_msg_writer() << tr("Balance: ") << print_money(m_wallet->balance()) << ", "
     << tr("unlocked balance: ") << print_money(m_wallet->unlocked_balance());
+  std::map<uint32_t, uint64_t> balance_per_subaddress = m_wallet->balance_per_subaddress();
+  std::map<uint32_t, uint64_t> unlocked_balance_per_subaddress = m_wallet->unlocked_balance_per_subaddress();
+  if (!detailed || balance_per_subaddress.empty())
+    return true;
+  success_msg_writer() << tr("Balance per address:");
+  success_msg_writer() << boost::format("%15s %21s %21s %21s %7s") % tr("Address") % tr("Balance") % tr("Unlocked balance") % tr("Label") % tr("Outputs");
+  std::vector<tools::wallet2::transfer_details> transfers;
+  m_wallet->get_transfers(transfers);
+  for (auto& i : balance_per_subaddress)
+  {
+    std::string address_str = m_wallet->get_subaddress_as_str(i.first).substr(0, 6);
+    cryptonote::subaddress_index subaddr_index = {m_wallet->get_current_subaddress_account(), i.first};
+    uint64_t num_unspent_outputs = std::count_if(transfers.begin(), transfers.end(), [&](const tools::wallet2::transfer_details& td) { return !td.m_spent && td.m_subaddr_index == subaddr_index; });
+    success_msg_writer() << boost::format(tr("%8u %6s %21s %21s %21s %7u")) % i.first % address_str % print_money(i.second) % print_money(unlocked_balance_per_subaddress[i.first]) % m_wallet->get_subaddress_label(i.first) % num_unspent_outputs;
+  }
   return true;
 }
 //----------------------------------------------------------------------------------------------------
 bool simple_wallet::show_balance(const std::vector<std::string>& args/* = std::vector<std::string>()*/)
 {
+  if (args.size() > 1 || (args.size() == 1 && args[0] != "detail"))
+  {
+    fail_msg_writer() << tr("usage: balance [detail]");
+    return true;
+  }
   LOCK_IDLE_SCOPE();
-  show_balance_unlocked();
+  show_balance_unlocked(args.size() == 1);
   return true;
 }
 //----------------------------------------------------------------------------------------------------
 bool simple_wallet::show_incoming_transfers(const std::vector<std::string>& args)
 {
+  if (args.size() > 2)
+  {
+    fail_msg_writer() << tr("usage: incoming_transfers [available|unavailable] [index=<N>]");
+    return true;
+  }
+  auto local_args = args;
   LOCK_IDLE_SCOPE();
 
   bool filter = false;
   bool available = false;
   bool verbose = false;
-  for (const auto &arg: args)
+  if (local_args.size() > 0)
   {
-    if (arg == "available")
+    if (local_args[0] == "available")
     {
       filter = true;
       available = true;
+      local_args.erase(local_args.begin());
     }
-    else if (arg == "unavailable")
+    else if (local_args[0] == "unavailable")
     {
       filter = true;
       available = false;
+      local_args.erase(local_args.begin());
     }
-    else if (arg == "verbose")
+    else if (local_args[0] == "verbose")
     {
       verbose = true;
+      local_args.erase(local_args.begin());
     }
   }
```

Comment:
OK. Actually I wasn't familiar with the `verbose` option since it wasn't documented in the help message. So I updated the help message as well.